### PR TITLE
ASAA-6 [Snippet Details Screen]

### PR DESCRIPTION
**Pull request by MMurrayAtBR**
Pull request for Snippet Details Screen, viewModel, widgets. Adds multiple api calls to request information about specific snippets.

### CODE REVIEW

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+import com.bottlerocketstudios.compose.snippets.SnippetUiModel
+import com.bottlerocketstudios.compose.snippets.snippetDetails.SnippetDetailsUiModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import org.koin.core.component.inject
+
+class SnippetDetailsViewModel : BaseViewModel() {
+    // DI
+    private val repo: BitbucketRepository by inject()
+
+    // State
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt]: Let's rename these "State" comments to "UI".  I usually reserve "State" for internal state that is not connected to UI.

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        childrenComments.forEach { child ->
+            childrenComments.find { it.id == child.parentId }?.childrenComments?.add(child) ?: run {
+                parentComments.find { it.id == child.parentId }?.childrenComments?.add(child)
+            }
+        }
+
+        snippetComments.value = parentComments.reversed()
+    }
+
+    /** Functions for optimized sort theory */
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt:165]: I'll see about getting this sort working as well as the original and then we can do a short presentation before a code n tell.

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsPrimaryButton.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import androidx.compose.ui.text.TextStyle
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.Dp
+import com.bottlerocketstudios.compose.resources.Colors
+import com.bottlerocketstudios.compose.resources.Dimens
+import com.bottlerocketstudios.compose.resources.typography
+import com.bottlerocketstudios.compose.util.Preview
+
+@Composable
+fun SnippetDetailsPrimaryButton(
```
Comment (BottleRocket-Colin) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsPrimaryButton.kt]: This could probably be name SnippetDetailsButton so as not to confuse it with the existing primary button composable

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/CircleAvatarImage.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.Dp
+import coil.compose.AsyncImage
+import com.bottlerocketstudios.compose.R
+import com.bottlerocketstudios.compose.resources.Dimens
+import com.bottlerocketstudios.compose.snippets.snippetDetails.returnMockSnippetDetails
+import com.bottlerocketstudios.compose.util.Preview
+
+@Composable
+fun CircleAvatarImage(
```
Comment (BottleRocket-Colin) [inline compose/src/main/java/com/bottlerocketstudios/compose/widgets/CircleAvatarImage.kt:27]: This could be a candidate to move into LaunchPad library,

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/UserCard.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.bottlerocketstudios.compose.widgets
+
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.tooling.preview.Preview
+
+@Composable
```
Comment (BottleRocket-Colin) [inline compose/src/main/java/com/bottlerocketstudios/compose/widgets/UserCard.kt]: Either document this liberally with TODOs for its intended purpose, or remove completely.

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/converter/LinksDtoConverter.kt [hunk 6]
```diff
@@ -5,6 +5,19 @@ import com.bottlerocketstudios.brarchitecture.data.model.LinksDto
 import com.bottlerocketstudios.brarchitecture.domain.models.Link
 import com.bottlerocketstudios.brarchitecture.domain.models.Links
 
-fun LinksDto.convertToLinks() = Links(avatar = avatar?.convertToLink())
+fun LinksDto.convertToLinks() = Links(
```
Comment (BottleRocket-Colin) [inline data/src/main/java/com/bottlerocketstudios/brarchitecture/data/converter/LinksDtoConverter.kt:8]: Good practice of isolating DTOs to data module.

**Review by BottleRocket-Colin — CHANGES_REQUESTED** [review]
Some minor changes in my comments.

File: .idea/.gitignore [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
+/shelf/
```
Comment (jaredstockton) [inline .idea/.gitignore:2]: @MMurrayAtBR Choose the most appropriate option below:

1. Delete this file completely if not truly needed OR
2. Delete this file and move the values to the primary project `.gitignore` inside the `### CUSTOM START` and `### CUSTOM END` section, taking the new path for each value into account (`.idea/shelf/`, `.idea/workspace.xml`). This will ensure that all of the .idea gitignore logic is located in one spot in the project.

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+import com.bottlerocketstudios.compose.snippets.snippetDetails.SnippetDetailsUiModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import org.koin.core.component.inject
+
+class SnippetDetailsViewModel : BaseViewModel() {
+    // DI
+    private val repo: BitbucketRepository by inject()
+
+    // State
+    private val workspaceId by lazy { MutableStateFlow("") }
```
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt]: @MMurrayAtBR I'm curious why you've made these lazy. What benefit do you see this bringing instead of just instantiating the properties immediately?

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+                getSnippetComments()
+                snippetDetailsDto.files?.keys?.toList()?.let { getRawFiles(it) }
+                isUserWatchingSnippet()
+            }
+        }
+    }
+
+    private fun getRawFiles(filePaths: List<String>) = launchIO {
+        snippetFiles.value = mutableListOf<SnippetDetailsFile?>().apply {
+            filePaths.forEach { path ->
```
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt]: @MMurrayAtBR Have you considered using `filePaths.map` instead of creating a mutableList, iterating with `filePaths.forEach`, and adding items to the list? I haven't tried this out for viability but think you should be able to get there. Let me know if you'd like more info or have questions :)

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+            }
+        }
+    }
+
+    /** Coded Response: Api returns 204 if user is watching and a 404 if user is not, else an error has occured */
+    private fun isUserWatchingSnippet() {
+        launchIO {
+            when (repo.isUserWatchingSnippet(workspaceId.value, encodedId.value)) {
+                is Status.Success -> isWatchingSnippet.value = true
+                is Status.Failure.Server -> isWatchingSnippet.value = false
```
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt]: @MMurrayAtBR Based on the kdoc function comment, let's check for a 404 value specifically here instead of the generic `Status.Failure.Server` value (which is any 4xx/5xx value) to behave properly when a non 404 server value is returned.

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        repo.deleteSnippetComment(workspaceId.value, encodedId.value, commentId).handlingErrors(R.string.delete_comment_error) { getSnippetComments() }
+    }
+
+    /////////////////////  Callbacks /////////////////////
+    fun changeSnippetWatching() = when (isWatchingSnippet.value) {
+        true -> stopWatchingSnippet()
+        false -> startWatchingSnippet()
+    }
+
+    fun onEditSnippetClick() {}
```
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt]: @MMurrayAtBR Add a TODO here mentioning this functionality needs to be added in the future

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsMockData.kt [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+import com.bottlerocketstudios.compose.util.asMutableState
+import java.time.ZonedDateTime
+import java.util.concurrent.TimeUnit
+import kotlin.math.roundToInt
+
+val mockCreationMsg = ZonedDateTime.now().minusDays(30).convertToTimeAgoMessage()
+val mockUpdatedMsg = ZonedDateTime.now().minusHours(18).convertToTimeAgoMessage()
+
+@Composable
+fun returnMockSnippetDetails() =
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsMockData.kt]: @MMurrayAtBR Make this `internal`

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsMockData.kt [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        elapsedHours in 1..23 -> "$elapsedHours Hours Ago"
+        elapsedDays in 1..6 -> "$elapsedDays Days Ago"
+        elapsedDays in 7..30 -> "${(elapsedDays / 7).toInt()} Weeks Ago"
+        elapsedDays in 31..365 -> "${(elapsedDays / 30.4).roundToInt()} Months Ago"
+        elapsedDays > 364 -> "${(elapsedDays / 365).toInt()} Years Ago"
+        else -> "Just Now"
+    }
+}
+
+val currentUser = User(
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsMockData.kt]: @MMurrayAtBR Make this `private` if possible, `internal` otherwise

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsMockData.kt [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+    username = null,
+    nickname = null,
+    displayName = "Luke Skywalker",
+    avatarUrl = "https://i.pinimg.com/736x/69/ed/be/69edbedeccf27136c2ea6b18af6ec49d.jpg",
+    accountStatus = null,
+    createdOn = null,
+    uuid = null,
+    links = null
+)
+val snippetComment = SnippetComment(
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsMockData.kt]: @MMurrayAtBR Make this `private` if possible, `internal` otherwise

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+    private val workspaceId by lazy { MutableStateFlow("") }
+    private val encodedId by lazy { MutableStateFlow("") }
+
+    // State - User
+    val currentUser = MutableStateFlow<User?>(null)
+    val isWatchingSnippet = MutableStateFlow(false)
+
+    // State - Snippet
+    val snippetDetails = MutableStateFlow<SnippetDetailsUiModel?>(null)
+    val snippetFiles = MutableStateFlow(mutableListOf<SnippetDetailsFile?>())
```
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt]: @MMurrayAtBR The signature could be improved by being a `MutableStateFlow<List<SnippetDetailsFile>>` instead of `MutableStateFlow<List<SnippetDetailsFile?>>` as there is no value in having nulls in the list.

1. Adjust this code
2. Adjust the `SnippetDetailsScreenState` (changing to `val files: State<List<SnippetDetailsFile>>`)
3. Adjust the code where new values are set in `snippetComments`.

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsUiModel.kt [hunk 16]
```diff
@@ -0,0 +1,19 @@
+package com.bottlerocketstudios.compose.snippets.snippetDetails
+
+import com.bottlerocketstudios.brarchitecture.domain.models.Links
+import com.bottlerocketstudios.brarchitecture.domain.models.SnippetDetailsFile
+import com.bottlerocketstudios.brarchitecture.domain.models.User
+
+data class SnippetDetailsUiModel(
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsUiModel.kt]: @MMurrayAtBR Have this class extend `DomainModel` (similar to `User` or `GitRepository`)

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsUiModel.kt [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+    val id: String? = null,
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsUiModel.kt]: @MMurrayAtBR When creating these Domain/Ui model classes, it's a good idea to "clean up" null types where possible to prevent unnecessary null handling from all consumers of this data.

For example, the following properties could be non-nullable Strings that use the empty String as the default value:

* id
* title
* createdMessage
* updatedMessage
* httpsCloneLink
* sshCloneLink

A null String or "" usually mean the same thing. If the type was `String?`, then you might do something like `foo.isNullOrEmpty()` or `foo.isBlankOrEmpty()`. If the type was `String`, then you'd just need to check `foo.isEmpty()` or `foo.isBlank()`

Additionally, if all of the other fields are required per the endpoint response, then they could be made non-null here in the domain/ui model.

cc: @BottleRocket-Colin

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsCommentItem.kt [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+                    onCancelClicked()
+                },
+            )
+        }
+    }
+}
+
+@Preview(showBackground = true)
+@Composable
+fun PreviewCommentCard() {
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsCommentItem.kt]: @MMurrayAtBR Make `private`

```suggestion
private fun PreviewCommentCard() {
```

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsCloneCard.kt [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+                singleLine = true,
+                textStyle = typography.h4,
+            )
+        }
+    }
+}
+
+@Preview(showBackground = true)
+@Composable
+fun PreviewCloneCard() {
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsCloneCard.kt]: @MMurrayAtBR Make `private`

```suggestion
private fun PreviewCloneCard() {
```

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsFileCard.kt [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+import com.bottlerocketstudios.compose.resources.brown_grey_three
+import com.bottlerocketstudios.compose.resources.typography
+import com.bottlerocketstudios.compose.snippets.snippetDetails.returnMockSnippetDetails
+import com.bottlerocketstudios.compose.util.convertToImageBitmap
+import com.bottlerocketstudios.compose.widgets.IconText
+import com.bottlerocketstudios.launchpad.compose.light
+import com.bottlerocketstudios.launchpad.compose.normal
+
+@Composable
+fun SnippetDetailsFilesCard(file: SnippetDetailsFile?) {
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsFileCard.kt]: @MMurrayAtBR If file is null, should anything render onscreen (I'm leaning towards it not)? I'd recommend making this argument non-null.

File: domain/src/main/java/com/bottlerocketstudios/brarchitecture/domain/models/SnippetDetailsFile.kt [hunk 21]
```diff
@@ -0,0 +1,7 @@
+package com.bottlerocketstudios.brarchitecture.domain.models
+
+data class SnippetDetailsFile(
```
Comment (jaredstockton) [inline domain/src/main/java/com/bottlerocketstudios/brarchitecture/domain/models/SnippetDetailsFile.kt:8]: @MMurrayAtBR Make this class extend `DomainModel`

File: domain/src/main/java/com/bottlerocketstudios/brarchitecture/domain/models/SnippetDetailsFile.kt [hunk 22]
```diff
@@ -0,0 +1,7 @@
...
+    val fileName: String?,
+    val links: Links?,
+    val rawFile: ByteArray? = null
```
Comment (jaredstockton) [inline domain/src/main/java/com/bottlerocketstudios/brarchitecture/domain/models/SnippetDetailsFile.kt]: @MMurrayAtBR Can this be null per the backend api call response schema? If not, let's make this a non-null type

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/network/BitbucketService.kt [hunk 23]
```diff
@@ -68,4 +73,82 @@ internal interface BitbucketService {
         @Part body: MultipartBody.Part,
         @Part("is_private") private: Boolean
     ): Response<SnippetDto>
+
+    /** https://developer.atlassian.com/cloud/bitbucket/rest/api-group-snippets/#api-snippets-workspace-encoded-id-delete */
```
Comment (jaredstockton) [inline data/src/main/java/com/bottlerocketstudios/brarchitecture/data/network/BitbucketService.kt:84]: @MMurrayAtBR Thanks for adding these links

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/ResponseToApiResultMapper.kt [hunk 24]
```diff
@@ -38,12 +41,22 @@ class ResponseToApiResultMapperImpl : ResponseToApiResultMapper {
         return when {
             response.isSuccessful -> Status.Success(Unit)
             else -> {
-                Timber.w("[toEmptyResult] Api not successful: message ${response.message()} code: ${response.code()}")
+                Timber.w("[toCodedResult] Api not successful: message ${response.message()} code: ${response.code()}")
```
Comment (jaredstockton) [inline data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/ResponseToApiResultMapper.kt]: @MMurrayAtBR Revert this change so that the logs use `toEmptyResult` (which matches the function name :)

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/converter/ZonedDateTimeConverter.kt [hunk 25]
```diff
@@ -0,0 +1,29 @@
+package com.bottlerocketstudios.brarchitecture.data.converter
+
+import java.time.ZonedDateTime
+import java.util.concurrent.TimeUnit
+import kotlin.math.roundToInt
+
+fun ZonedDateTime.convertToTimeAgoMessage(): String {
```
Comment (jaredstockton) [inline data/src/main/java/com/bottlerocketstudios/brarchitecture/data/converter/ZonedDateTimeConverter.kt]: @MMurrayAtBR Since `compose` and `data` both declare this `ZonedDateTime.convertToTimeAgoMessage()` function (copy/pasted from one to the other), delete this code and move this to the `domain` module, which both `compose` and `data` modules depend on.

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsMockData.kt [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+        onCommentChanged =  {},
+        onSaveCommentClick = {},
+        onCancelNewCommentClick = {},
+        onDeleteCommentClick = {},
+        onEditCommentClick = {},
+        newReplyComment = "".asMutableState(),
+        onReplyChanged = {}
+    )
+
+fun ZonedDateTime.convertToTimeAgoMessage(): String {
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/SnippetDetailsMockData.kt]: @MMurrayAtBR Since `compose` and `data` both declare this `ZonedDateTime.convertToTimeAgoMessage()` function (copy/pasted from one to the other), delete this code and move this to the `domain` module, which both `compose` and `data` modules depend on.

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/converter/SnippetDetailsDtoConverter.kt [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+        creator = creator?.convertToUser(),
+        links = links?.convertToLinks(),
+        httpsCloneLink = links?.convertToLinks()?.clone?.find { it?.name == "https" }?.href,
+        sshCloneLink = links?.convertToLinks()?.clone?.find { it?.name == "ssh" }?.href,
+    )
+
+fun Map<String, LinksDto?>?.convertToUiModel(): List<SnippetDetailsFile?>? =
+    this?.map { SnippetDetailsFile(it.key, it.value?.convertToLinks()) }
+
+
```
Comment (jaredstockton) [inline data/src/main/java/com/bottlerocketstudios/brarchitecture/data/converter/SnippetDetailsDtoConverter.kt]: @MMurrayAtBR Run `ktlintFormat` to remove these additional blank lines.

File: data/build.gradle.kts [hunk 28]
```diff
@@ -106,6 +106,7 @@ class ApiKeyProperties(pathToProperties: String, project: Project) {
 
 dependencies {
     implementation(project(mapOf("path" to ":domain")))
+    implementation(project(mapOf("path" to ":compose")))
```
Comment (jaredstockton) [inline data/build.gradle.kts]: @MMurrayAtBR This line should be removed as `data` should not be dependent on `compose`. What code lives in `compose` that is needed in `data`?

cc: @BottleRocket-Colin

File: compose/src/main/res/values/strings.xml [hunk 29]
```diff
@@ -59,6 +59,30 @@
     </plurals>
     <string name="today">Today</string>
 
+    <!-- Snippet Details -->
+    <string name="created_by">Created By %1s</string>
```
Comment (jaredstockton) [inline compose/src/main/res/values/strings.xml]: @MMurrayAtBR Append string name with `_format` suffix to make it clear to the consumer of this string that it should require format args :)

File: compose/src/main/res/values/strings.xml [hunk 30]
```diff
@@ -59,6 +59,30 @@
...
+    <string name="owned_by">Owned By</string>
+    <string name="last_modified">Last Modified</string>
+    <string name="header_snippet_files">Snippet Files</string>
+    <string name="header_snippet_comments">Snippet Comments %1s</string>
```
Comment (jaredstockton) [inline compose/src/main/res/values/strings.xml]: @MMurrayAtBR Append string name with `_format` suffix to make it clear to the consumer of this string that it should require format args :)

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/CircleAvatarImage.kt [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+            .height(sizeDp)
+            .clip(CircleShape),
+        placeholder = painterResource(R.drawable.ic_avatar_placeholder),
+        contentScale = ContentScale.Crop,
+    )
+}
+
+@Preview(showBackground = true)
+@Composable
+fun PreviewCircleAvatarImage() {
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/widgets/CircleAvatarImage.kt]: @MMurrayAtBR Make `private`

```suggestion
private fun PreviewCircleAvatarImage() {
```

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/CircleAvatarImage.kt (hunk 4 above)
Comment (jaredstockton) [reply compose/src/main/java/com/bottlerocketstudios/compose/widgets/CircleAvatarImage.kt:27]: @BottleRocket-Colin I like the idea but think that before doing that it would be nice to make some things adjustable such as:

* border size (dps)
* border color?
* contentDescription?

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsPrimaryButton.kt [hunk 32]
```diff
@@ -0,0 +1,61 @@
...
+                text = buttonText,
+                style = it,
+            )
+        }
+    }
+}
+
+@Preview(showBackground = true)
+@Composable
+fun PreviewSnippetDetailsPrimaryButton() {
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsPrimaryButton.kt]: @MMurrayAtBR Make `private`

```suggestion
private fun PreviewSnippetDetailsPrimaryButton() {
```

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsFileCard.kt [hunk 33]
```diff
@@ -0,0 +1,123 @@
...
+                    .padding(Dimens.grid_2)
+                    .fillMaxWidth()
+            )
+        }
+    }
+}
+
+@Preview(showSystemUi = true)
+@Composable
+fun PreviewFilesCard() {
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsFileCard.kt]: @MMurrayAtBR Make `private`

```suggestion
private fun PreviewFilesCard() {
```

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsFileCard.kt [hunk 34]
```diff
@@ -0,0 +1,123 @@
...
+                        .fillMaxWidth()
+                )
+                RawFileData(byteArray = file?.rawFile)
+            }
+        }
+    }
+}
+
+@Composable
+fun RawFileData(byteArray: ByteArray?) {
```
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsFileCard.kt]: @MMurrayAtBR If byteArry is null, nothing is rendered onscreen. I'd recommend making this argument non-null and  not calling the function instead of taking in a nullable type and doing nothing

cc: @BottleRocket-Colin

**Review by jaredstockton — CHANGES_REQUESTED** [review]
@MMurrayAtBR Nice work! Please take a look at the comments and address them. Afterwards, I'll take another look :)

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/CircleAvatarImage.kt (hunk 4 above)
Comment (BottleRocket-Colin) [reply compose/src/main/java/com/bottlerocketstudios/compose/widgets/CircleAvatarImage.kt:27]: Indeed

Comment (MMurrayAtBR) [reply compose/src/main/java/com/bottlerocketstudios/compose/widgets/CircleAvatarImage.kt:27]: @br-jared-stockton that's a great idea. I've added those in my update.

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/UserCard.kt (hunk 5 above)
Comment (MMurrayAtBR) [reply compose/src/main/java/com/bottlerocketstudios/compose/widgets/UserCard.kt]: Should have been removed; went with a different idea.

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt (hunk 8 above)
Comment (MMurrayAtBR) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt]: @br-jared-stockton Oversight. I had originally thought these values may not always be used, but I should have updated them once I knew the direction I was going.

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsFileCard.kt (hunk 34 above)
Comment (MMurrayAtBR) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/snippetDetails/snippetDetailsWidgets/SnippetDetailsFileCard.kt]: Good idea

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt [hunk 35]
```diff
@@ -0,0 +1,195 @@
...
+
+    private fun startWatchingSnippet() = launchIO {
+        repo.startWatchingSnippet(workspaceId.value, encodedId.value).handlingErrors(R.string.error_changing_watching) { isUserWatchingSnippet() }
+    }
+
+    // TODO: Show dialog to confirm user wants to continue with deletion before calling this function
+    fun onDeleteSnippetClick() = launchIO {
+        repo.deleteSnippet(workspaceId.value, encodedId.value).handlingErrors(R.string.delete_snippet_error) {
+            // Using handleError here to inform the user that the snippet has been deleted; maybe use a branded messaging?
+            handleError(R.string.delete_snippet_success)
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetDetailsViewModel.kt]: Let's create a notify function that is similar to handle error

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/repository/BitbucketRepository.kt [hunk 36]
```diff
@@ -146,11 +161,80 @@ internal class BitbucketRepositoryImpl(
...
     }
 
+    override suspend fun deleteSnippet(workspaceId: String, encodedId: String): Status<Unit> {
+        return wrapRepoExceptions("deleteSnippet") {
+            bitbucketService.deleteSnippet(workspaceId, encodedId).toEmptyResult()
+        }
+    }
+
+    override suspend fun getSnippetDetails(workspaceId: String, encodedId: String): Status<SnippetDetailsDto> {
+        return wrapRepoExceptions("getSnippetDetails") {
```
Comment (BottleRocket-Colin) [inline data/src/main/java/com/bottlerocketstudios/brarchitecture/data/repository/BitbucketRepository.kt]: could this, and others, be converted to single expression functions.

**Review by BottleRocket-Colin — CHANGES_REQUESTED** [review]
Just a couple of minor comments from me.

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/ResponseToApiResultMapper.kt (hunk 24 above)
Comment (MMurrayAtBR) [reply data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/ResponseToApiResultMapper.kt]: Oh no! Glad you caught that

File: data/build.gradle.kts (hunk 28 above)
Comment (MMurrayAtBR) [reply data/build.gradle.kts]: Hmm, I don't recall adding that. It's also in :app build.gradle, but the git blame annotation is Br* ?
Honestly have no idea why that it's there. Could I have triggered some kind of dependecy adding function in AS?

Comment (jaredstockton) [reply data/build.gradle.kts]: @MMurrayAtBR It should be present in the `:app` module (representing the entire Android application) since the `:app` module contains the Activities and relies on the :compose module to provide the composables for rendering.

The `:data` module should not be aware of UI and therefore shouldn't need to also depend on `:compose`, since it shouldn't be using or relying on any code from that module.

Not sure how it got added but please remove just here :)

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/BaseViewModel.kt [hunk 37]
```diff
@@ -56,6 +56,15 @@ abstract class BaseViewModel : ViewModel(), KoinComponent {
         }
     }
 
+    /**
+     * Used to display a notification message with standard UI pattern
+     */
+    suspend fun notifyUser(@StringRes messageId: Int) {
```
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/BaseViewModel.kt:62]: Thank you!

**Review by BottleRocket-Colin — APPROVED** [review]

**Review by BottleRocket-Colin — APPROVED** [review]
